data/DataTrans.py

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+
+import pandas as pd
+import numpy as np
+import csv
+Max_Value=3000
+Min_Value=0
+
+def get_Zone(Z_csv):
+    zone=pd.read_csv(Z_csv)
+
+    # Create dictionary to store the mapping of states and codes
+    zone_dic={}
+    for i in range(len(zone)):
+        zone_dic[zone['STATE'][i]]=zone['ID'][i]
+    return zone_dic
+
+
+# Clean data; throw substations which are outside the United States and not available.
+def Clean(E_csv,zone_dic):
+    csv_data = pd.read_csv(E_csv)
+    Num_sub=len(csv_data)
+    row_indexs=[]
+    for i in range(Num_sub):
+        if((csv_data['STATE'][i] not in zone_dic) or (csv_data['STATUS'][i] != 'IN SERVICE') ):
+            row_indexs.append(i)
+    clean_data=csv_data.drop(labels=row_indexs)
+    return clean_data
+
+
+
+#Create dictionary to store neighbors of each sub
+def Neighbors(T_csv):
+    Trans_data = pd.read_csv(T_csv, usecols=['VOLTAGE','SUB_1','SUB_2'])
+    N_dict={}
+    for i in range(len(Trans_data)):
+        if(Trans_data['SUB_1'][i] in N_dict):
+            N_dict[Trans_data['SUB_1'][i]].append(Trans_data['SUB_2'][i])
+        else:
+            N_dict[Trans_data['SUB_1'][i]]=[Trans_data['SUB_2'][i]]
+        if(Trans_data['SUB_2'][i] in N_dict):
+            N_dict[Trans_data['SUB_2'][i]].append(Trans_data['SUB_1'][i])
+        else:
+            N_dict[Trans_data['SUB_2'][i]]=[Trans_data['SUB_1'][i]]
+    return N_dict
+
+#Calculate the base_KV for each node
+
+def cal_KV(clean_data,N_dict):
+    KV_dict={}   # store base_KV of each node
+    to_cal=[]  # store nodes needed to be calculated
+   
+    for index, row in clean_data.iterrows():
+        base_KV=-999999.0
+        Min_Vol=row['MIN_VOLT']
+        Max_Vol=row['MAX_VOLT']
+        if(Min_Vol>=Max_Value or Min_Vol<=Min_Value): # only values between 0 to 3000 are valid
+            if(Max_Vol<Max_Value and Max_Vol>Min_Value):
+                base_KV=Max_Vol
+            else:
+                to_cal.append(row['NAME'])
+                continue
+        else:
+            if(Max_Vol<Max_Value and Max_Vol>Min_Value):
+                base_KV=(Max_Vol+Min_Vol)/2
+            else:
+                base_KV=Min_Vol
+        KV_dict[row['NAME']]=base_KV
+
+     #Go through to_cal repeatedly until it doesn;t change  
+    length=100000
+    while(len(to_cal)<length):
```

Comment:
As I am debugging on this code block, it runs several iterations and not able to populate some specific substations during all the iterations. The loop stops when we hit len(to_cal)==length. And the size of to_cal at the end is roughly 6000+. I suspect these are islands and we should probably remove them ahead of this algo to avoid hitting them again and again, with no way to populate them.

Reply:
Thanks for your advice. Let me explain why I use such a "inefficient" code.

There are two kinds of specific substation: Some substations are totally isolated, not connected to any others. They're easily to be found. However, there're some specific substation groups. 2 or 3 substations are connected with each other, all of them are "dirty", form a "dirty" subgraph.

At present, I use neighbors of a "dirty" substation to estimate its base_KV value( 9000+ nodes are estimated by this way) . I assume every node will be converted to "clean" node finally. Thus, it's important to decide which nodes should be discarded. For example , if a node has an "estimated" neighbor, should I keep it or throw it when doing estimation?

Definitely, I can go through the graph and find all maximal connected subgraph. Then calculate them one by one. If the dataset is large enough, this will lower time complexity.

Reply:
But I did make a mistake. I should skip the nodes which have no connections. Thank you very much.

Reply:
The new algorithm is to running a neighbor-search and calculating the average of the KV. It is more efficient now. Resolving this.